File: src/contests/cg/puzzle/scala/mirrors.scala
// 🎮 CodinGame Puzzle - mirrors
// https://www.codingame.com/training/easy/mirrors

object Solution {
  def main(args: Array[String]): Unit = {
    scala.io.StdIn.readLine()
    val r = scala.io.StdIn.readLine().trim.split(" ").map(_.toDouble)
    var reflected = 0.0
    for (i <- r.indices.reverse) {
      val ri = r(i)
      val denom = 1 - ri * reflected
      reflected = ri + (if (denom == 0) 0.0 else ((1 - ri) * (1 - ri) * reflected) / denom)
    }
    println("%.4f".formatLocal(java.util.Locale.US, reflected))
  }
}

File: src/contests/cg/puzzle/perl/mirrors.pl
# 🎮 CodinGame Puzzle - mirrors
# https://www.codingame.com/training/easy/mirrors

<STDIN>;
my @r = split ' ', <STDIN>;
my $reflected = 0.0;
for (my $i = $#r; $i >= 0; $i--) {
    my $ri = $r[$i];
    my $denom = 1 - $ri * $reflected;
    $reflected = $ri + ($denom == 0 ? 0 : ((1 - $ri) * (1 - $ri) * $reflected) / $denom);
}
printf "%.4f\n", $reflected;
